### PR TITLE
[istio-authz-kfam] Support Istio Authorization Policy in KFAM

### LINKED ISSUE #5399: [istio-authz-kfam] Support Istio Authorization Policy in KFAM
/kind feature

Umbrella issue: https://github.com/kubeflow/kubeflow/issues/5313

This may or may not be implemented, depending on whether we will manage to deprecate KFAM before the 1.3 release.

**Why you need this feature:**

In order to use the `kfam` with Istio >= 1.6
- We use Istio authorization to limit access to network endpoints, like Jupyter Notebooks.
- Early Istio used Istio RBAC.
- Newer Istio deprecated Istio RBAC and moved to Istio AuthorizationPolicy.

**Describe the solution you'd like:**

1. Add an envvar option `ISTIO_AUTHZ_TYPE`, with possible values of `rbac`(default) and `authorizationpolicy`.
2. Implement the `authorizationpolicy` option.

This way we also ensure we don't break people that can't move to the new CRD yet.

/cc @kubeflow/wg-notebook-leads

### PR DESCRIPTION
**Pull request by Subreptivus**
Closes #5399
Umbrella issue: #5313

### CODE REVIEW

**Comment by Subreptivus** [comment]
@googlebot I signed it!

**Comment by yanniszark** [comment]
@Subreptivus thanks a ton for the contribution, this is sorely needed :)
A general comment I have is that I believe we want to support both Istio RBAC and Istio AuthorizationPolicy.
This is because partial rollout of AuthorizationPolicy is difficult. If Istio detects any RBAC resources, it will use the RBAC model, not AuthorizationPolicy. So rollout has to be done once all components support it. Do you think you can refactor the PR a bit so that the type of Istio resources applied is configurable based on an environment variable? For example, it can be called `ISTIO_AUTHZ_TYPE` with possible values of `rbac` (default) and `authorizationpolicy`.

What do you think?

**Comment by thesuperzapper** [comment]
We really need to decide what approach we are taking with Istio 1.6 (and more generally in kubeflow) before we implement any specific fixes.

Also, since this component is owned by the Notebook WG, we should probably discuss this at the next Notebook WG meeting @kubeflow/wg-notebook-leads

**Comment by Subreptivus** [comment]
Hey @yanniszark!
I beg your pardon for late answer, yet I'll agree with @thesuperzapper here. I can spend some time to make it work using your approach, but I would rather wait for final decision about future plan of integration new Istio versions. The thing here, that this isn't only about KFAM, but a lot more components.
For instance I've integrated another PR (#5393) in this regard into our custom Kubeflow installation, yet I had to extend it, because it breaking KF/KNative. There is KFP components should be attended as well. There is not only components code, but manifests should be changed as well.
So I'll keep an eye on this thread and others that I've found regarding new Istio versions. Hope you guys come up with some good plan for the further steps and I could help you with that.

**Comment by akirillov** [comment]
Hi all. I’m currently working on a similar issue and wondering why `ServiceRole` is managed in both the Profile Controller and KFAM? Shouldn’t this be handled only by the Profile Controller as a part of `Profile` reconciliation?

**Comment by yanniszark** [comment]
@Subreptivus thanks a lot for your effort on this! Given the total deprecation of Istio RBAC and the reduced amount of work necessary to maintain Istio AuthorizationPolicy only, I agree that we should move forward with upgrading to AuthorizationPolicy only and removing Istio RBAC.

I'm going to review this as part of wg-manifests. Is the PR ready for review?

**Comment by Subreptivus** [comment]
@yanniszark 
I guess so, apart from `go.mod` (this is always driving me nuts, and I'm just make it buildable without really checking what is going on there).
And yes, I'm still interested in making it all compatible with Istio v1.6+, so you can count on me if I can help somehow...

**Comment by yanniszark** [comment]
@Subreptivus that's great! IIUC with this PR, KFAM stops creating Istio RBAC resources and creates Istio AuthorizationPolicy resources instead. If the PR is ready for review, let me know by replying here and I will review it asap.

**Comment by Subreptivus** [comment]
@yanniszark It's ready.

File: components/access-management/kfam/api_default.go
Comment (DavidSpek) [inline components/access-management/kfam/api_default.go:50]: Is there a reason for double spaces here?

**Comment by DavidSpek** [comment]
I know about 0 Go, but for what it is worth it all looks good and makes sense to me. Only thing I saw was some double spaces that I don't believe are necessary.

File: components/access-management/kfam/api_default.go
Comment (Subreptivus) [reply components/access-management/kfam/api_default.go:50]: I don't know, my believe that go linter did that change.

Comment (filintod) [reply components/access-management/kfam/api_default.go:50]: [gofmt](https://golang.org/cmd/gofmt/) would align them, and that is usually used in go projects

**Review by yanniszark — CHANGES_REQUESTED** [review]
@subreptivus thanks for the PR!
I only have a couple of comments. 
I have prepared a branch with proposed changes, if you agree please push them here:
https://github.com/arrikto/kubeflow/tree/kfam_istio_1_6_compatible-rev

File: components/access-management/kfam/bindings.go
Comment (Subreptivus) [reply components/access-management/kfam/bindings.go:60]: I'm not aware of all configurations, but in ours Azure AD we're getting uppercased username by default in User Principal Name (UPN) claim. So if we will try to add contributor to some namespace, then the name of the binding for the user `TEST@example.com` would be `user-example-com-clusterrole-edit` instead of `user-test-example-com-clusterrole-edit`

File: components/access-management/kfam/bindings.go
Comment (Subreptivus) [reply components/access-management/kfam/bindings.go:79]: By default Istio flavored Kubeflow installation, there is anonymous user injection to any request which doesn't have UserID header set (https://github.com/kubeflow/manifests/blob/v1.2-branch/istio/add-anonymous-user-filter/base/kustomization.yaml). We aren't installing this filter, and we're explicitly denying request without UserID header.
And this code should cover it. According to [documentation](https://pkg.go.dev/istio.io/api/security/v1beta1#AuthorizationPolicy), not defining `Rules` in AuthorizationPolicy is equivalent to setting a default of deny for the target workloads.

Comment (yanniszark) [reply components/access-management/kfam/bindings.go:79]: @subreptivus in this PR, we're just trying to port the existing functionality to AuthorizationPolicy.
Allowing all requests to a namespace if a header is not provided to KFAM is new behavior, a dangerous one too!
If you think this is needed, let's discuss in an issue. But for this PR, let's scope it down to just port what we already have to AuthorizationPolicy.

File: components/access-management/kfam/bindings.go
Comment (yanniszark) [reply components/access-management/kfam/bindings.go:60]: I would suggest we leave this change out of the PR and scope it down to just porting what we already have to AuthorizationPolicy. If you need this new feature, let's discuss on a new issue.

**Comment by yanniszark** [comment]
@Subreptivus sorry for the delay and thanks for your responses! My general feedback is to try and keep this PR scoped to just porting what we already have to AuthorizationPolicy. For subsequent changes, we can discuss in new issues.

File: components/access-management/kfam/bindings.go
Comment (Subreptivus) [reply components/access-management/kfam/bindings.go:60]: It looks very minor change for new issue, but ok, you're the boss here, I'll create new one.

File: components/access-management/kfam/bindings.go
Comment (Subreptivus) [reply components/access-management/kfam/bindings.go:79]: Yeah, you're right. I've just checked.
First of all, there is check for the header/argument name, rather than actual value of the header. My bad, I've missed it, it wasn't my initial intention.
And second, you're right to. I thought if I'll add empty condition, it will be like following:
```yaml
spec:
 rules:
 - {}
 ```
 But in fact it produces the following:
 ```yaml
spec:
  rules:
  - when: []
 ```
 which is really not what I wanted to achieve. So this is settled.

**Comment by Subreptivus** [comment]
@yanniszark So what are my next steps? I'm asking because all the changes from here are already in https://github.com/arrikto/kubeflow/tree/kfam_istio_1_6_compatible-rev
Should just close this one?

**Comment by yanniszark** [comment]
@Subreptivus if you're ok with the changes, you can push them in this PR to continue!
```sh
# On the PR branch
git remote add arrikto git@github.com:arrikto/kubeflow.git
git fetch --all
git reset --hard arrikto/kfam_istio_1_6_compatible-rev
git push
```

**Comment by yanniszark** [comment]
@Subreptivus you should also rebased ontop of latest master, as @DavidSpek mentioned:
```sh
# On PR branch
git fetch --all
git rebase origin/master
```

**Comment by DavidSpek** [comment]
Just wanted to inform you that this PR will need to be rebased on master again. This is because the CI/CD has changes slightly and for prow to pick it up the configs need to be present in the PR. Sorry for the inconvenience.

**Comment by DavidSpek** [comment]
Here is the log output from the build process:
```
go: finding github.com/google/martian v2.1.0+incompatible
go: finding honnef.co/go/tools v0.0.0-20190106161140-3f1c8253044a
go: finding golang.org/x/text v0.3.1-0.20180807135948-17ff2d5776d2
go: finding github.com/googleapis/gax-go/v2 v2.0.4
go: finding google.golang.org/genproto v0.0.0-20190418145605-e7d98fc518a7
go: finding github.com/google/pprof v0.0.0-20181206194817-3ea8567a2e57
INFO[0063] Taking snapshot of full filesystem...
INFO[0082] RUN if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -gcflags 'all=-N -l' -o
 access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'all=-N -l' -o access-managemen
t main.go;     fi
INFO[0082] cmd: /bin/sh
INFO[0082] args: [-c if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -gcflags 'all=-N
-l' -o access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'all=-N -l' -o access-man
agement main.go;     fi]
INFO[0082] Running: [/bin/sh -c if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -gcfla
gs 'all=-N -l' -o access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'all=-N -l' -o
 access-management main.go;     fi]
# github.com/kubeflow/kubeflow/components/access-management/kfam
kfam/bindings.go:47:18: undefined: Binding
kfam/bindings.go:48:18: undefined: Binding
kfam/model_binding_list.go:13:6: BindingEntries redeclared in this block
        previous declaration at kfam/model_binding.go:13:6
kfam/model_binding_list.go:15:13: undefined: Binding
error building image: error building stage: failed to execute command: waiting for process to exit: exit status 2
```

**Comment by yanniszark** [comment]
@Subreptivus sorry for the constant changes. There's a lot of activity going on right now in CI. Could you rebase to pick up the latest CI and fix the conflict?

**Comment by DavidSpek** [comment]
@yanniszark Is this PR ready to merge after the rebase? If so, I would like to suggest merging it quickly so it doesn't need to be rebased another time (or hold PRs related to KFAM until after this is merged). 

@Subreptivus Sorry for all the inconvenience. I'm pretty sure it was a PR of mine that caused this problem.

**Comment by DavidSpek** [comment]
```
INFO[0057] Taking snapshot of full filesystem...
INFO[0074] RUN if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -gcflags 'all=-N
-l' -o access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'all=-N -l' -o acce
ss-management main.go;     fi
INFO[0074] cmd: /bin/sh
INFO[0074] args: [-c if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -gcflags 'a
ll=-N -l' -o access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'all=-N -l' -
o access-management main.go;     fi]
INFO[0074] Running: [/bin/sh -c if [ "$(uname -m)" = "aarch64" ]; then         CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build
-gcflags 'all=-N -l' -o access-management main.go;     else         CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -gcflags 'a
ll=-N -l' -o access-management main.go;     fi]
# github.com/kubeflow/kubeflow/components/access-management/kfam
kfam/api_default.go:72:2: roleBindingLister declared and not used
error building image: error building stage: failed to execute command: waiting for process to exit: exit status 2
```

It seems like the rebase from master didn't go properly.

**Comment by DavidSpek** [comment]
You can follow the test that is running now [here](https://argo.kubeflow-testing.com/workflows/kubeflow-test-infra/kubeflow-kubeflow-presubmit-ac-mgr-tests-5439-30e21f1-6320-ac4d)

**Comment by Subreptivus** [comment]
@DavidSpek Can you just use arrikto's branch and close this one already?
https://github.com/arrikto/kubeflow/tree/kfam_istio_1_6_compatible-rev

**Comment by DavidSpek** [comment]
@Subreptivus Sadly not my call to make, @yanniszark will need to make that decision.

**Comment by DavidSpek** [comment]
@Subreptivus I recreated the PR. Can you reply `@googlebot I consent` on this PR? https://github.com/kubeflow/kubeflow/pull/5725

**Comment by DavidSpek** [comment]
@Subreptivus Thank you for all the effort and sorry for all the difficulties with rebasing that caused this PR to not be merged. The Arrikto branch was just merged into the repo as you requested so I am closing this PR. 

/close

**Pull request closed without merging**
